src/wccp/WccpConfig.cc

```diff
@@ -367,342 +288,339 @@ operator==(const char *text, ts::ConstBuffer const &b)
+    if (portType) {
+      std::string text = portType.as<std::string>();
+      if ("src" == text) {
+        svc_info.enableFlags(ServiceGroup::PORTS_SOURCE);
+      } else if ("dst" == text) {
+        svc_info.disableFlags(ServiceGroup::PORTS_SOURCE);
       } else {
-        zret.push(Prop_Invalid_Type(prop, ts::config::StringValue));
+        zret.push(Port_Type_Invalid(text, portType.Mark().line));
       }
     }
 
     // Ports for service.
     svc_info.clearPorts();
-    if ((prop = svc_cfg[SVC_PROP_PORTS]).hasValue()) {
-      src_line = prop.getSourceLine();
+
+    YAML::Node ports = svc_cfg[SVC_PROP_PORTS];
+    if (ports) {
       if (ServiceGroup::STANDARD == svc_info.getSvcType()) {
-        zret.push(Svc_Prop_Ignored_In_Standard(SVC_PROP_PORTS, src_line));
+        zret.push(Svc_Prop_Ignored_In_Standard(SVC_PROP_PORTS, ports.Mark().line));
       } else {
-        if (prop.isContainer()) {
-          size_t nport = prop.childCount();
-          size_t pidx, sidx;
-          bool malformed_error = false;
+        if (ports.IsSequence()) {
+          size_t nport = ports.size();
+          size_t sidx  = 0;
+
           // Clip to maximum protocol allowed ports.
           if (nport > ServiceGroup::N_PORTS) {
-            zret.push(Svc_Ports_Too_Many(src_line, nport));
+            zret.push(Svc_Ports_Too_Many(ports.Mark().line, nport));
             nport = ServiceGroup::N_PORTS;
           }
+
           // Step through the ports.
-          for (pidx = sidx = 0; pidx < nport; ++pidx) {
-            Value const &port_cfg = prop[pidx];
-            if (ts::config::IntegerValue == port_cfg.getType()) {
-              x = atoi(port_cfg.getText()._ptr);
-              if (0 <= x && x <= 65535)
-                svc_info.setPort(sidx++, x);
-              else
-                zret.push(Svc_Prop_Out_Of_Range(SVC_PROP_PORTS, port_cfg, x, 0, 65535));
-            } else if (!malformed_error) { // only report this once.
-              zret.push(Svc_Ports_Malformed(src_line));
-              malformed_error = true;
+          for (auto it = ports.begin(); it != ports.end(); ++it) {
+            const YAML::Node &port_cfg = *it;
+            int x                      = port_cfg.as<int>();
+            if (0 <= x && x <= 65535) {
+              svc_info.setPort(sidx++, x);
+            } else {
+              zret.push(Svc_Prop_Out_Of_Range(SVC_PROP_PORTS, ports.Mark().line, x, 0, 65535));
             }
           }
-          if (sidx)
+
+          if (sidx) {
             svc_info.enableFlags(ServiceGroup::PORTS_DEFINED);
-          else
-            zret.push(Svc_Ports_None_Valid(src_line));
+          } else {
+            zret.push(Svc_Ports_None_Valid(ports.Mark().line));
+          }
         } else {
-          zret.push(Prop_Invalid_Type(prop, ts::config::ListValue));
+          // port is a scalar
+          int x = ports.as<int>();
+          svc_info.setPort(0, x); // ??? start at 0? same with sidx
```

Comment:
Heh. You could move the `sidx` declaration to just under the `else` and use `svc_info.setPort(sidx++, x);` here as well.